### ISSUE TITLE
Micro-Benchmark of (ill-advised) string concatenation in a loop

### ISSUE THREAD

**Issue by mingodad** [LuaJIT/LuaJIT#939]
Doing some tests with the code bellow I found that `LuaJit` is slower than normal `Lua`, probably there is a hanging fruit here.
```
/usr/bin/time luajit joinStrings.lua 
strPlusStr took	840000	6.587768
strToArrayPush took	840000	0.0010059999999994
2.24user 4.34system 0:06.59elapsed 99%CPU (0avgtext+0avgdata 16520maxresident)k
0inputs+0outputs (0major+2806095minor)pagefaults 0swaps
```
```
/usr/bin/time lua-5.1 joinStrings.lua 
strPlusStr took	840000	3.133744
strToArrayPush took	840000	0.0098340000000001
1.83user 1.30system 0:03.14elapsed 99%CPU (0avgtext+0avgdata 18856maxresident)k
0inputs+0outputs (0major+758416minor)pagefaults 0swaps
```
```
/usr/bin/time lua-5.3 joinStrings.lua 
strPlusStr took	840000	2.836371
strToArrayPush took	840000	0.0096310000000002
1.26user 1.58system 0:02.84elapsed 99%CPU (0avgtext+0avgdata 8444maxresident)k
0inputs+0outputs (0major+1126903minor)pagefaults 0swaps
```
```
/usr/bin/time lua-5.4 joinStrings.lua 
strPlusStr took	840000	3.46929
strToArrayPush took	840000	0.005941
1.39user 2.08system 0:03.48elapsed 100%CPU (0avgtext+0avgdata 7040maxresident)k
0inputs+0outputs (0major+1449269minor)pagefaults 0swaps
```
```lua
local str1 = "abc";
local str2 = "defghijklmnñopqrstuvwxyz";

function strPlusStr(times)
  local str = "";
  for i =1,times do
    str = str .. str1 ..  str2;
  end
  return str;
end

local tblin = table.insert;

function strToArrayPush(times)
  local str = {};
  for i=1,times do
    tblin(str, str1);
    tblin(str, str2);
  end
  return table.concat(str);
end

function test(times)
  local time, str;

  time = os.clock();
  str = strPlusStr(times);
  print("strPlusStr took", #str, os.clock()-time)

  time = os.clock();
  str = strToArrayPush(times);
  print("strToArrayPush took", #str, os.clock()-time)
end

test(30000);
```

**Comment by MikePall** [comment]
First, that's not a particularly interesting micro-benchmark. Concatenating strings in a loop is known to be a bad idea in Lua and other languages with immutable strings.

And, as the second part of it shows, LuaJIT is 10x faster than Lua 5.1 when the proper language idiom `table.concat()` is used. I'm sure using LuaJIT's string.buffer would be even faster.

That said, the differences are due to a combination of memory fragmentation, cache effects, string interning and HashDoS mitigations.

One could try to sink the intermediate `IR_BUFSTR`, but that's not so easy in the general case.

Sure, it would make this particular micro-benchmark 10000x faster. But for most real-world programs a concatenation result is either used immediately afterwards -or- unused and easily eliminated (like in `for i=1,100 do assert(i, "X"..i) end`).

TL;DR: I doubt there's an enhancement waiting here, which is both easy to implement *and* useful in practice.
